lib/AST/GenericSignatureBuilder.cpp

```diff
@@ -3679,6 +3619,49 @@ PotentialArchetype *GenericSignatureBuilder::realizePotentialArchetype(
   return pa;
 }
 
+static Type substituteConcreteType(GenericSignatureBuilder &builder,
+                                   PotentialArchetype *basePA,
+                                   TypeDecl *concreteDecl) {
+  if (!concreteDecl)
+    return Type();
+
+  auto *proto =
+      concreteDecl->getDeclContext()->getAsProtocolOrProtocolExtensionContext();
+
+  if (!concreteDecl->hasInterfaceType())
+    builder.getLazyResolver()->resolveDeclSignature(concreteDecl);
+
+  if (concreteDecl->hasInterfaceType()) {
```

Comment:
Nit: I suggest doing an early return if `!concreteDecl->hasInterfaceType()` to de-nest the main code path.

Reply:
Sure! I confess that I have a habit of moving the code around without really refactoring it all the way...